trieste/acquisition/sampler.py

```diff
@@ -186,358 +174,55 @@ def binary_search(val: float) -> float:  # Find empirical interquartile range
-        # matrix of feature evaluations across the training data :math:`\Phi` and
-        # observation noise variance :math:`\sigma^2`.
-
-        phi = self._feature_functions(self._dataset.query_points)  # [n, m]
-        G = tf.matmul(phi, phi, transpose_b=True)  # [n, n]
-        s = self._noise_variance * tf.eye(self._num_data, dtype=phi.dtype)
-        L = tf.linalg.cholesky(G + s)
-        L_inv_phi = tf.linalg.triangular_solve(L, phi)  # [n, m]
-        L_inv_y = tf.linalg.triangular_solve(L, self._dataset.observations)  # [n, 1]
-
-        theta_posterior_mean = tf.tensordot(tf.transpose(L_inv_phi), L_inv_y, [[-1], [-2]])[
-            :, 0
-        ]  # [m,]
-        theta_posterior_covariance = tf.eye(self._num_features, dtype=phi.dtype) - tf.tensordot(
-            tf.transpose(L_inv_phi), L_inv_phi, [[-1], [-2]]
-        )  # [m, m]
-        theta_posterior_chol_covariance = tf.linalg.cholesky(theta_posterior_covariance)  # [m, m]
-
-        return tfp.distributions.MultivariateNormalTriL(
-            theta_posterior_mean, theta_posterior_chol_covariance
-        )
-
-    def get_trajectory(self) -> TrajectoryFunction:
-        """
-        Generate an approximate function draw (trajectory) by sampling weights
-        and evaluating the feature functions.
-
-        :return: A trajectory function representing an approximate trajectory from the Gaussian
-            process, taking an input of shape `[N, D]` and returning shape `[N, 1]`
-        """
-
-        theta_sample = self._theta_posterior.sample(1)  # [1, m]
-
-        def trajectory(x: TensorType) -> TensorType:
-            feature_evaluations = self._feature_functions(x)  # [N, m]
-            return tf.matmul(feature_evaluations, theta_sample, transpose_b=True)  # [N,1]
-
-        return trajectory
-
-    def sample(self, at: TensorType) -> TensorType:
-        """
-        Return approximate samples from either the objective function's minimser or its minimal
-        value over the candidate set `at`.
-
-        :param at: Where to sample the predictive distribution, with shape `[N, D]`, for points
-            of dimension `D`.
-        :return: The samples, of shape `[S, D]` (where `S` is the `sample_size`) if sampling
-            the function's minimser or shape `[S, 1]` if sampling the function's mimimal value.
-        :raise ValueError (or InvalidArgumentError): If ``at`` has an invalid shape.
-        """
-        tf.debugging.assert_shapes([(at, ["N", None])])
-
         if self._sample_min_value:
             thompson_samples = tf.zeros([0, 1], dtype=at.dtype)  # [0,1]
         else:
             thompson_samples = tf.zeros([0, tf.shape(at)[1]], dtype=at.dtype)  # [0,D]
 
-        for _ in tf.range(self._sample_size):
-            sampled_trajectory = self.get_trajectory()
+        for _ in tf.range(sample_size):
+            sampled_trajectory = trajectory_sampler.get_trajectory()
             evaluated_trajectory = sampled_trajectory(at)  # [N, 1]
```

Comment:
why do we first create an instance and then pass `at`? it seems to me we could simplify things and pass it to `get_trajectory` method and get evaluation from there?

Reply:
In the future we will want to optimize sampled_trajectory using a gradient optimizer instead of just picking the best over a discrete grid (as happens here). Therefore, we need to be able to return the whole trajecotry